Multi-platform docker image build

danieldabate opened the pull request:
Hi!
Currently the Docker image only supports linux/amd64 arch. 

This pull request adds multi-platform support for the platforms that Home Assistant currently supports:
- linux/386
- linux/amd64
- linux/arm/v6
- linux/arm/v7
- linux/arm64/v8

Tested generating an image on a separate github repo and running it on my RPI4. Validated that the image runs well now, but I have not tested the RTSPtoWeb functionality itself.

allenporter commented:
Awesome, thank you for doing this.
Does this change the output names of the build images?
Do you have an output of the workflow build log? I am curious to check it out, but did not see it on the github fork.

allenporter commented on .github/workflows/publish-docker.yml:
```diff
@@ -36,6 +42,7 @@ jobs:
         uses: docker/build-push-action@v2.9.0
         with:
           context: .
+          platforms: linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64/v8
```
I sometimes see folks do this with `strategy` and `matrix`. Does that have any advantages like running in parallel compared to doing this way? I am not familiar with the build-push-action detials.

danieldabate replied:
Not an expert either, but I believe the outcome is the main difference
between both approaches and they tackle two different needs.

In my approach only one Docker image is generated with all the platforms,
allowing for easier distribution. Because of this, the name of the image
and tag is not changed from the current ones. Building for the different
platforms is done in parallel, but in the same runner.

With the matrix approach you'd get one image per platform. This of course
generates smaller images, but each user needs to pull the specific tag for
their platform, making distribution a bit harder. Building for the
different platforms is done in parallel, I think in different runners.

Both approaches I believe should work together. Most of the users first
download the multi platform build and might then switch to the platform
specific image if want to have a smaller image.

I implemented only the first one as it is the one I know how to implement
and it’s the less “disruptive”. I can try implementing also the matrix
approach, but can’t promise anything 😅

danieldabate replied:
Well.. after doing some tests I found that when pulling a multi-platform image, only the part of the image corresponding to the specific host's platform is pulled. The other platforms don't get pulled.

This means that the matrix approach wouldn't bring any improvements over the multi-platform docker image approach for the end user. Only (maybe) multi runner builds, which I don't believe is necessary for this project as the multi-image approach still builds in parallel pretty fast, as you can see in the build logs.

You can find the build logs of the latest test build here:
https://github.com/danieldabate/RTSPtoWeb/actions/runs/1875816937

And the multi-platform Docker image here: 
https://github.com/danieldabate/RTSPtoWeb/pkgs/container/rtsptoweb/15439851?tag=v2.2.0

If you do:
`docker pull ghcr.io/danieldabate/rtsptoweb:v2.2.0`
you will see that the hash of the pulled image is the same as if you pull only the piece of you host's platform _(you can find those under the **OS / Arch** tab)_.

allenporter replied:
Thanks for the detail. It sounds like the only win could be that it runs the actions in parallel, but it just takes 3 minutes, so not too bad at all.

Image on a raspberry pi:
```
$ docker image ls ghcr.io/danieldabate/rtsptoweb
REPOSITORY                       TAG       IMAGE ID       CREATED       SIZE
ghcr.io/danieldabate/rtsptoweb   latest    f5d4e70cd318   5 hours ago   37.6MB
ghcr.io/danieldabate/rtsptoweb   v2.2.0    f5d4e70cd318   5 hours ago   37.6MB
```

Image on amd64:
```
$ docker image ls ghcr.io/danieldabate/rtsptoweb
REPOSITORY                       TAG       IMAGE ID       CREATED       SIZE
ghcr.io/danieldabate/rtsptoweb   latest    b7554063b244   5 hours ago   40.7MB
ghcr.io/danieldabate/rtsptoweb   v2.2.0    b7554063b244   5 hours ago   40.7MB
```

Great!

danieldabate commented:
Not an expert either, but I believe the outcome is the main difference
between both approaches and they tackle two different needs.

In my approach only one Docker image is generated with all the platforms,
allowing for easier distribution. Because of this, the name of the image
and tag is not changed from the current ones. Building for the different
platforms is done in parallel, but in the same runner.

With the matrix approach you'd get one image per platform. This of course
generates smaller images, but each user needs to pull the specific tag for
their platform, making distribution a bit harder. Building for the
different platforms is done in parallel, I think in different runners.

Both approaches I believe should work together. Most of the users first
download the multi platform build and might then switch to the platform
specific image if want to have a smaller image.

I implemented only the first one as it is the one I know how to implement
and it’s the less “disruptive”. I can try implementing also the matrix
approach, but can’t promise anything 😅

Also, I deleted and cleaned the packages and releases on my repo to avoid
confusion (and OCD 😂). I'll re run them tomorrow and share the outcome
with you.

On Mon, 21 Feb 2022 at 04:38 Allen Porter ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/publish-docker.yml
> <https://github.com/deepch/RTSPtoWeb/pull/105#discussion_r810756223>:
>
> > @@ -36,6 +42,7 @@ jobs:
>          uses: ***@***.***
>          with:
>            context: .
> +          platforms: linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64/v8
>
> I sometimes see folks do this with strategy and matrix. Does that have
> any advantages like running in parallel compared to doing this way? I am
> not familiar with the build-push-action details.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/deepch/RTSPtoWeb/pull/105#pullrequestreview-888170817>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP4BOKSN2OEFBK4H2Q27M3U4GXT3ANCNFSM5O5LM4AQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

allenporter commented:
Cheers @danieldabate 👍🏼

allenporter commented:
I'm releasing 2.2.0 to rebuild images with your improvements here: https://github.com/deepch/RTSPtoWeb/actions/runs/1877288860

danieldabate commented:
Awesome!! Thanks!
